### PR TITLE
Separate out VM dependencies

### LINKED ISSUE #17: Remove 'dart:io' dependencies where not strictly needed
dart2js must be able to compile _itself_ so it cannot use 'dart:io'. This means I can't use the discovery mechanism in this repo, but I would still like to use the `Packages` class. We should move things around a little to separate out the 'dart:io' dependent code from the platform agnostic code.

### LINKED ISSUE #1507: Remove 'dart:io' dependencies where not strictly needed
dart2js must be able to compile _itself_ so it cannot use 'dart:io'. This means I can't use the discovery mechanism in this repo, but I would still like to use the `Packages` class. We should move things around a little to separate out the 'dart:io' dependent code from the platform agnostic code.

### PR DESCRIPTION
**Pull request by harryterkelsen**
Fixes dart-lang/tools#1507 
cc @pq @sethladd @lrhn

### CODE REVIEW

**Comment by lrhn** [comment]
What is the purpose of this change?

File: lib/packages.dart [hunk 1]
```diff
@@ -34,34 +32,6 @@ abstract class Packages {
   factory Packages(Map<String, Uri> packageMapping) =>
       new MapPackages(packageMapping);
```
Comment (lrhn) [inline lib/packages.dart]: How about removing this constructor as well, and just not import any implementations. I'd be fine with that.
If someone wants to be clever, they can import src/packages_impl directly (and we could move it out of src).

File: lib/src/packages_impl_html.dart [hunk 2]
```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library package_config.packages_impl_html;
```
Comment (sethladd) [inline lib/src/packages_impl_html.dart]: does this library require dart:html ?

File: lib/src/packages_impl_vm.dart [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  Uri getBase(String packageName) =>
+    new Uri.file(path.join(_packageDir.path, packageName, '.'));
+
+  Iterable<String> _listPackageNames() {
+    return _packageDir.listSync()
+    .where((e) => e is Directory)
+    .map((e) => path.basename(e.path));
+  }
+
+  Iterable<String> get packages {
```
Comment (sethladd) [inline lib/src/packages_impl_vm.dart]: `... packages => _listPackageNames()` ?

Comment (harryterkelsen) [inline lib/src/packages_impl_vm.dart]: done

File: lib/packages.dart (hunk 1 above)
Comment (harryterkelsen) [inline lib/packages.dart]: done

**Comment by harryterkelsen** [comment]
Responded to comments, PTAL

**Comment by harryterkelsen** [comment]
Oops, fixed the tests

File: lib/src/packages_impl_html.dart (hunk 2 above)
Comment (harryterkelsen) [inline lib/src/packages_impl_html.dart]: no. I have renamed it to just 'packages_impl'

**Comment by lrhn** [comment]
It's fine now. We can still fiddle with the separation and whether we want the constructor to be there, but what we have now is ok.
I'll squash the commits and land them (I prefer using git-cl because it does that for you).

**Pull request closed without merging**